sdk_contrib/async_processor/src/main/java/io/opentelemetry/sdk/contrib/trace/export/DisruptorEventQueue.java

```diff
@@ -102,73 +102,82 @@ public DisruptorEvent newInstance() {
     this.blocking = blocking;
   }
 
-  void enqueue(ReadableSpan readableSpan, EventType eventType) {
-    enqueue(readableSpan, eventType, null);
-  }
-
-  // Enqueues an event on the {@link DisruptorEventQueue}.
-  void enqueue(ReadableSpan readableSpan, EventType eventType, CountDownLatch flushLatch) {
+  void enqueueStartEvent(ReadableSpan span) {
     if (isShutdown) {
       if (!loggedShutdownMessage.getAndSet(true)) {
         logger.info("Attempted to enqueue entry after Disruptor shutdown.");
       }
       return;
     }
+    enqueue(EventType.ON_START, span, null);
+  }
 
-    if (blocking) {
-      ringBuffer.publishEvent(TRANSLATOR_THREE_ARG, eventType, readableSpan, flushLatch);
-    } else {
-      // TODO: Record metrics if element not added.
-      ringBuffer.tryPublishEvent(TRANSLATOR_THREE_ARG, eventType, readableSpan, flushLatch);
+  void enqueueEndEvent(ReadableSpan span) {
+    if (isShutdown) {
+      if (!loggedShutdownMessage.getAndSet(true)) {
+        logger.info("Attempted to enqueue entry after Disruptor shutdown.");
+      }
+      return;
     }
+    enqueue(EventType.ON_END, span, null);
   }
 
-  // Shuts down the underlying disruptor.
+  // Shuts down the underlying disruptor. Ensures that when this method returns the disruptor is
+  // shutdown.
   void shutdown() {
-    enqueueAndLock(EventType.ON_SHUTDOWN);
+    synchronized (this) {
```

Comment:
synchronizing on `this` is potentially dangerous, as you don't have control over who else might try to grab your monitor and keep this from ever executing.

Reply:
Customized the messages to be more relevant.

Reply:
This is an internal class so I would not worry that much about that.

Reply:
@jkwatson are you ok with my comment?

Reply:
yeah, I'm not convinced that it's a pattern we want to perpetuate, but it's ok for now.